std::span::operator[] raises "span index out of range" error in ComplexTextController::adjustGlyphsAndAdvances

fujii opened the pull request:
#### 5757471893c37c07f97e55cbff426a90ab07cf49
<pre>
std::span::operator[] raises &quot;span index out of range&quot; error in ComplexTextController::adjustGlyphsAndAdvances
<a href="https://bugs.webkit.org/show_bug.cgi?id=286905">https://bugs.webkit.org/show_bug.cgi?id=286905</a>

Reviewed by Darin Adler.

With Windows Debug CRT, an expression &quot;charactersSpan[characterIndex +
1]&quot; in ComplexTextController::adjustGlyphsAndAdvances raised &quot;span
index out of range&quot; error for fast/text/word-break-letter-spacing-utf16-surrogates.html.

Before accessing the element, we have to ensure
`charactersSpan.size()` is enough.

* Source/WebCore/platform/graphics/ComplexTextController.cpp:
(WebCore::ComplexTextController::adjustGlyphsAndAdvances):

Canonical link: <a href="https://commits.webkit.org/289794@main">https://commits.webkit.org/289794@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1438a8a8e7ebc820d7d83b47080abbc2005266e2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87792 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/7308 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/42179 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92657 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/38542 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/89843 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/7689 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/15479 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67790 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25531 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90794 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5876 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79436 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48158 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5663 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33842 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37649 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/76061 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34723 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/94544 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14960 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/11001 "Found 1 new test failure: ipc/create-media-source-with-invalid-constraints-crash.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76639 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/15215 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75292 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75875 "Found 1 new API test failure: /WebKitGTK/TestWebKitPolicyClient:/webkit/WebKitPolicyClient/new-window-policy (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20241 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18676 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/7962 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13734 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14976 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/20279 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14720 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/18164 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/16502 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/platform/graphics/ComplexTextController.cpp:
```diff
@@ -757,7 +757,8 @@ void ComplexTextController::adjustGlyphsAndAdvances()
                 bool isFirstCharacter = !(characterIndex + complexTextRun.stringLocation());
-                bool isLastCharacter = characterIndexInRun + 1 == m_run.length() || (U16_IS_LEAD(character) && characterIndexInRun + 2 == m_run.length() && U16_IS_TRAIL(charactersSpan[characterIndex + 1]));
+                bool isLastCharacter = characterIndexInRun + 1 == m_run.length()
+                    || (U16_IS_LEAD(character) && characterIndexInRun + 2 == m_run.length() && characterIndex + 2 == charactersSpan.size() && U16_IS_TRAIL(charactersSpan[characterIndex + 1]));
```
Why `characterIndex + 2 == charactersSpan.size()` rather than `characterIndex + 2 >= charactersSpan.size()`?

fujii replied:
`>=` is also fine. This condition is checking if this is the last character of m_run. m_run is split into some m_complexTextRuns. For example, assuming m_run "abc😀" is split into "abc" and "😀". If "😀" is the last character of m_run, both conditions `characterIndexInRun + 2 == m_run.length()` and `characterIndex + 2 == charactersSpan.size()` are true.

darinadler replied:
There’s a guarantee that no characters are after `m_run` in `charactersSpan` in the “is last character” case? Not even things like whitespace?

fujii replied:
I don't understand your question. For example, assuming m_run "abc😀😎xyz" is split into "abc", "😀😎" and "xyz". And, "😎" is the current character, this is not the last character because "xyz" follows. The condition `characterIndexInRun + 2 == m_run.length()` is false.

darinadler replied:
Never mind. I thought that maybe `m_run` was sometimes not at the very end of the span but was the "last" but I guess that is not the case.

fujii committed:
std::span::operator[] raises "span index out of range" error in ComplexTextController::adjustGlyphsAndAdvances
https://bugs.webkit.org/show_bug.cgi?id=286905

Reviewed by Darin Adler.

With Windows Debug CRT, an expression "charactersSpan[characterIndex +
1]" in ComplexTextController::adjustGlyphsAndAdvances raised "span
index out of range" error for fast/text/word-break-letter-spacing-utf16-surrogates.html.

Before accessing the element, we have to ensure
`charactersSpan.size()` is enough.

* Source/WebCore/platform/graphics/ComplexTextController.cpp:
(WebCore::ComplexTextController::adjustGlyphsAndAdvances):

Canonical link: https://commits.webkit.org/289794@main
